client/app/components/Parameters.jsx

```diff
@@ -69,14 +84,15 @@ export class Parameters extends React.Component {
 
   setPendingValue = (param, value, isDirty) => {
     const { onPendingValuesChange } = this.props;
-    this.setState(({ parameters }) => {
+    this.setState(({ parameters, touched }) => {
       if (isDirty) {
         param.setPendingValue(value);
+        touched[param.name] = true;
```

Comment:
Shouldn't `touched` be cloned before mutating?

Reply:
Seems needless to me.
I wouldn't consider this a side-effect since it's being done in the context of `setState` and since this flag isn't designed to "toggle off" but only "on", there's no risk of it being out-of-sync.

Reply:
It's not about side effects, rather about change detection. You return the same `touched` object - how React knows that something changed? Even if it works now - I think it's accidental; what if one day React will optimize `setState` and it will no longer update component if you return the same object from `setState` callback? IMHO it worth fixing, especially considering that fix is the single line of code: `touched = { ...touched, [param.name]: true };`